File: 113_PathSumII.cpp
// Topic: Path Sum II
// Description: Find all root-to-leaf paths where sum equals target.
// Approach: DFS backtracking collect paths.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTreeLevel(const vector<int>& vals)
{
    if (vals.empty())
        return nullptr;
    int n = vals.size();
    if (vals[0] == INT_MIN)
        return nullptr;
    vector<TreeNode*> nodes(n, nullptr);
    for (int i = 0; i < n; ++i)
        if (vals[i] != INT_MIN)
            nodes[i] = new TreeNode(vals[i]);
    for (int i = 0; i < n; ++i)
        if (nodes[i])
        {
            int l = 2 * i + 1, r = 2 * i + 2;
            if (l < n)
                nodes[i]->left = nodes[l];
            if (r < n)
                nodes[i]->right = nodes[r];
        }
    return nodes[0];
}

void dfs(TreeNode* node, int target, vector<int>& cur, vector<vector<int>>& res)
{
    if (!node)
        return;
    cur.push_back(node->val);
    if (!node->left && !node->right && node->val == target)
        res.push_back(cur);
    else
    {
        dfs(node->left, target - node->val, cur, res);
        dfs(node->right, target - node->val, cur, res);
    }
    cur.pop_back();
}

vector<vector<int>> pathSum(TreeNode* root, int target)
{
    vector<vector<int>> res;
    vector<int> cur;
    dfs(root, target, cur, res);
    return res;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<pair<vector<int>, int>> tests = {{{5, 4, 8, 11, -1, 13, 4, 7, 2, -1, -1, -1, 1}, 22},
                                            {{1, 2, 3}, 5}};

    for (int t = 0; t < (int)tests.size(); ++t)
    {
        auto vals = tests[t].first;
        int target = tests[t].second;
        TreeNode* root = buildTreeLevel(vals);
        auto res = pathSum(root, target);
        cout << "Test " << (t + 1) << ":\n";
        if (res.empty())
            cout << "(no paths)\n";
        for (auto& p : res)
        {
            for (size_t i = 0; i < p.size(); ++i) cout << p[i] << (i + 1 == p.size() ? '\n' : ' ');
        }
    }
    return 0;
}

/*
Time Complexity: O(n) to visit nodes; number of paths may cause extra cost
Space Complexity: O(h) recursion + output
*/

File: 113_PathSumII.py
# Topic: Path Sum II
# Description: Find all root-to-leaf paths where sum equals target.
# Approach: DFS backtracking to collect paths.

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree_level(vals: List[int]) -> Optional[TreeNode]:
    if not vals or vals[0] is None:
        return None
    nodes = [TreeNode(v) if v is not None else None for v in vals]
    for i, node in enumerate(nodes):
        if node:
            l, r = 2 * i + 1, 2 * i + 2
            if l < len(nodes):
                node.left = nodes[l]
            if r < len(nodes):
                node.right = nodes[r]
    return nodes[0]


class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        result = []
        path = []

        def dfs(node: Optional[TreeNode], remaining: int) -> None:
            if not node:
                return
            path.append(node.val)
            if not node.left and not node.right and node.val == remaining:
                result.append(path[:])
            else:
                dfs(node.left, remaining - node.val)
                dfs(node.right, remaining - node.val)
            path.pop()

        dfs(root, targetSum)
        return result


if __name__ == "__main__":
    tests = [([5, 4, 8, 11, None, 13, 4, 7, 2, None, None, 5, 1], 22),
             ([1, 2, 3], 5)]
    for i, (vals, target) in enumerate(tests):
        root = build_tree_level(vals)
        print(f"Test {i + 1}: {Solution().pathSum(root, target)}")

"""
Time Complexity: O(n) to visit nodes; number of paths may cause extra cost
Space Complexity: O(h) recursion + output
"""
